arima.py

```diff
@@ -0,0 +1,198 @@
+
+    res_train = pd.DataFrame(res[0:train_size])
+    res_val = pd.DataFrame(res[train_size:res.shape[0]])
+
+    res_train_2 = res_train.dropna()
+
+    X_train = res_train_2.iloc[:, 1:].values.reshape(-1, q)
+    y_train = res_train_2.iloc[:, 0].values.reshape(-1, 1)
+
+    model = sm.OLS(y_train, sm.add_constant(X_train)).fit()
+
+    theta = np.array(model.params[1:])
+    intercept = model.params[0]
+
+    res_train_2['Predicted_Values'] = X_train.dot(theta) + intercept
+    # res_train_2[['Residuals','Predicted_Values']].plot()
+
+    X_val = res_val.iloc[:, 1:].values.reshape(-1, q)
+    res_val['Predicted_Values'] = X_val.dot(theta) + intercept
+    # res_val[['Residuals', 'Predicted_Values']].plot()
+
+    from sklearn.metrics import mean_squared_error
+    RMSE = np.sqrt(mean_squared_error(
+        res_val['Residuals'], res_val['Predicted_Values']))
+    AIC = model.aic
+    BIC = model.bic
+
+    # print("The RMSE is:", RMSE, ", Value of AIC is:", AIC,
+    #       ", Value of BIC is:", BIC, ", Value of q:", q)
+    return [res_train_2, res_val, theta, intercept, RMSE, AIC, BIC]
+
+
+def opt_p(df_testing, low_p, up_p, str, key):
+    ''' Pick the best p based on lowest aic or bic or RMSE as specified'''
+    assert (str == 'AIC' or str == 'BIC' or str == 'RMSE')
+    best_p_aic, best_p_bic, best_p_rmse = 0, 0, 0
+    best_rmse, best_aic, best_bic = float("inf"), float("inf"), float("inf")
+    for p in range(low_p, up_p):
+        df_train_2, df_val, theta, intercept, RMSE, AIC, BIC = AR(
+            p, pd.DataFrame(df_testing), key)
+        if RMSE < best_rmse:
+            best_p_rmse = p
+            best_rmse = RMSE
+        if AIC < best_aic:
+            best_p_aic = p
+            best_aic = AIC
+        if BIC < best_bic:
+            best_p_bic = p
+            best_bic = BIC
+    if (str == 'RMSE'):
+        return best_p_rmse, best_rmse
+    elif (str == 'AIC'):
+        return best_p_aic, best_aic
+    elif (str == 'BIC'):
+        return best_p_bic, best_bic
+
+
+def opt_q(df_testing, low_q, up_q, str):
+    ''' Pick the best q based on lowest aic or bic or RMSE as specified'''
+    assert (str == 'AIC' or str == 'BIC' or str == 'RMSE')
+    best_q_aic, best_q_bic, best_q_rmse = 0, 0, 0
+    best_rmse, best_aic, best_bic = float("inf"), float("inf"), float("inf")
+    for q in range(low_q, up_q):
+        res_train_2, res_val, theta, intercept, RMSE, AIC, BIC = MA(
+            q, pd.DataFrame(df_testing))
+        if RMSE < best_rmse:
+            best_q_rmse = q
+            best_rmse = RMSE
+        if AIC < best_aic:
+            best_q_aic = q
+            best_aic = AIC
+        if BIC < best_bic:
+            best_q_bic = q
+            best_bic = BIC
+    if (str == 'RMSE'):
+        return best_q_rmse, best_rmse
+    elif (str == 'AIC'):
+        return best_q_aic, best_aic
+    elif (str == 'BIC'):
+        return best_q_bic, best_bic
+
+
+def arima(low_p, high_p, low_q, high_q, df, key, opt_method):
```

Comment:
Consider overall documenting the meaning of each parameter as well! The specs are super concise and awesome with this exception. Would probs help lots with other people who would want to use this code